ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/SSLEngineUnwrapper.kt

```diff
@@ -0,0 +1,107 @@
+
+    fun cancel(cause: Throwable?) {
+        input.cancel(cause)
+    }
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    suspend inline fun readAndUnwrap(
+        initialUnwrapDestination: ByteBuffer,
+        updateUnwrapDestination: (ByteBuffer) -> Unit
+    ): SSLEngineResult? {
+        if (!unwrapLock.tryLock()) return suspendCancellableCoroutine {
+            synchronized(this) {
+                unwrapResultCont = it
+            }
+        }
+        try {
+            var unwrapDestination: ByteBuffer = initialUnwrapDestination
+            var result: SSLEngineResult?
+
+            if (unwrapRemaining > 0) {
+                unwrapSource.compact()
+                unwrapSource.flip()
+            } else {
+                unwrapSource.clear()
+                if (!readData()) {
+                    resumeUnwrapContinuation(null)
+                    return null
+                }
+            }
+
+            while (true) {
+                result = engine.unwrap(unwrapSource, unwrapDestination)
+
+                when (result.status!!) {
+                    SSLEngineResult.Status.BUFFER_UNDERFLOW -> {
+                        if (unwrapSource.limit() == unwrapSource.capacity()) {
+                            // buffer is too small to read all needed data
+                            unwrapSource = bufferAllocator.reallocatePacket(unwrapSource, flip = false)
+                        } else {
+                            // not all data received
+                            unwrapSource.position(unwrapSource.limit())
+                            unwrapSource.limit(unwrapSource.capacity())
+                        }
+                        if (!readData()) {
+                            resumeUnwrapContinuation(null)
+                            return null
+                        }
+                    }
+                    SSLEngineResult.Status.BUFFER_OVERFLOW -> {
+                        unwrapDestination = bufferAllocator.reallocateApplication(unwrapDestination, flip = true)
+                    }
+                    else -> break
+                }
+            }
+            unwrapRemaining = unwrapSource.remaining()
+            if (unwrapDestination !== initialUnwrapDestination) updateUnwrapDestination(unwrapDestination)
+            resumeUnwrapContinuation(result)
+            return result
+        } catch (cause: Throwable) {
+            synchronized(this) {
+                unwrapResultCont?.resumeWithException(cause)
+                unwrapResultCont = null
+            }
+            throw cause
+        } finally {
+            unwrapLock.unlock()
+        }
+    }
+
+    private fun resumeUnwrapContinuation(result: SSLEngineResult?) {
+        synchronized(this) {
+            unwrapResultCont?.resume(result)
+            unwrapResultCont = null
+        }
+    }
+
+    private suspend fun readData(): Boolean {
+        val read = input.readAvailable(unwrapSource)
```

Comment:
Could we feed SSLEngine directly from the channel without a buffer? (method providing a block).

We also should be careful if the channel gets canceled with an exception.

Reply:
I would try to split up several extensions on ByteReadChannel and ByteWriteChannel feeding SSLEngine with the available data

Reply:
> SSLEngine directly from the channel without a buffer

Not really, as SSLEngine can say, that it needs more data to unwrap data, but buffer from channel f.e. is of some fixed size, while SSLEngine need bigger buffer (BUFFER_UNDERFLOW case)